OpenEdXMobile/src/main/java/org/edx/mobile/view/BulkDownloadFragment.java

```diff
@@ -0,0 +1,428 @@
+package org.edx.mobile.view;
+
+import android.annotation.SuppressLint;
+import android.content.Context;
+import android.databinding.DataBindingUtil;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.annotation.NonNull;
+import android.support.annotation.StringRes;
+import android.support.v4.view.ViewCompat;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.CompoundButton;
+
+import com.joanzapata.iconify.Icon;
+import com.joanzapata.iconify.fonts.FontAwesomeIcons;
+import com.joanzapata.iconify.internal.Animation;
+
+import org.edx.mobile.R;
+import org.edx.mobile.base.BaseFragment;
+import org.edx.mobile.core.IEdxEnvironment;
+import org.edx.mobile.databinding.RowBulkDownloadBinding;
+import org.edx.mobile.logger.Logger;
+import org.edx.mobile.model.VideoModel;
+import org.edx.mobile.model.course.CourseComponent;
+import org.edx.mobile.model.course.HasDownloadEntry;
+import org.edx.mobile.model.course.VideoBlockModel;
+import org.edx.mobile.model.db.DownloadEntry;
+import org.edx.mobile.model.download.NativeDownloadModel;
+import org.edx.mobile.module.db.DataCallback;
+import org.edx.mobile.module.prefs.VideoPrefs;
+import org.edx.mobile.module.storage.DownloadInterruptedEvent;
+import org.edx.mobile.module.storage.DownloadStartedEvent;
+import org.edx.mobile.util.MemoryUtil;
+import org.edx.mobile.util.NetworkUtil;
+import org.edx.mobile.util.ResourceUtil;
+import org.edx.mobile.view.adapters.NewCourseOutlineAdapter;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import de.greenrobot.event.EventBus;
+
+public class BulkDownloadFragment extends BaseFragment {
+    protected final Logger logger = new Logger(getClass().getName());
+
+    public enum SwitchState {ON, OFF, DEFAULT}
+
+    private static final int DELETE_DELAY_MS = 4000;
+    private static final int DOWNLOAD_PROGRESS_DELAY_MS = 2000;
+
+    private RowBulkDownloadBinding binding;
+    private NewCourseOutlineAdapter.DownloadListener downloadListener;
+    private IEdxEnvironment environment;
+    private VideoPrefs prefManager;
+    private SwitchState switchState = SwitchState.DEFAULT;
+    private CourseComponent courseComponent;
+    private Timer downloadProgressTimer;
```

Comment:
According to android docs: https://developer.android.com/reference/java/util/Timer.html

`Java 5.0 introduced the java.util.concurrent package and one of the concurrency utilities therein is the ScheduledThreadPoolExecutor which is a thread pool for repeatedly executing tasks at a given rate or delay. It is effectively a more versatile replacement for the Timer/TimerTask combination, as it allows multiple service threads, accepts various time units, and doesn't require subclassing TimerTask (just implement Runnable). Configuring ScheduledThreadPoolExecutor with one thread makes it equivalent to Timer.`

We should avoid to use Timer class and can use Handler or other Android framework api for scheduling tasks.

Reply:
Hmmm, I would definitely consider this as an improvement now that we have used TimerTask quite extensively for this feature.

Reply:
This didn't turn out to be much of an effort and improved the app's performance noticeably. The change has been made in this commit fce177e.

Reply:
Great work! ⭐️ 🌟